File: src/carpcomm/db/recordfile.py
#!/usr/bin/python

# Author: Timothy Stranex <tstranex@carpcomm.com>
# Copyright 2013 Timothy Stranex

#!/usr/bin/python

"""Simple binary record file format."""

import struct

RECORD_WRITER_V0_HEADER = 'RecordWriter000'


class RecordWriter(object):
    def __init__(self, f):
        self.f = f
        self.f.write(RECORD_WRITER_V0_HEADER)

    def Write(self, record):
        self.f.write(struct.pack('!L', len(record)))
        self.f.write(record)

    def Close(self):
        self.f.close()


class RecordReader(object):
    def __init__(self, f):
        self.f = f
        h = f.read(len(RECORD_WRITER_V0_HEADER))
        assert h == RECORD_WRITER_V0_HEADER

    def Read(self):
        s = self.f.read(4)
        if not s:
            return None
        size = struct.unpack('!L', s)[0]
        return self.f.read(size)

File: src/carpcomm/db/recordfile.go
// Author: Timothy Stranex <tstranex@carpcomm.com>
// Copyright 2013 Timothy Stranex

// Simple binary record file format.

package db

import "io"
import "errors"
import "log"
import "os"

const kRecordWriterV0Header = "RecordWriter000"

type RecordReader struct {
	r io.Reader
}

func NewRecordReader(r io.Reader) (*RecordReader, error) {
	header := make([]byte, len(kRecordWriterV0Header))
	_, err := io.ReadFull(r, header)
	if err != nil {
		return nil, err
	}
	if string(header) != kRecordWriterV0Header {
		return nil, errors.New("Invalid RecordWriter header.")
	}

	return &RecordReader{r}, nil
}

func NewRecordReaderForFile(path string) (*RecordReader, error) {
	f, err := os.Open(path)
	if err != nil {
		log.Printf("File open error: %s", err.Error())
		return nil, err
	}
	rr, err := NewRecordReader(f)
	if err != nil {
		log.Printf("RecordReader error: %s", err.Error())
		return nil, err
	}
	return rr, nil
}

func (rr *RecordReader) ReadRecord() ([]byte, error) {
	size := make([]byte, 4)
	_, err := io.ReadFull(rr.r, size)
	if err != nil {
		return nil, err
	}

	// Big-endian
	var n uint32
	for i := 0; i < 4; i++ {
		n = (n << 8) | (uint32)(size[i])
	}

	rec := make([]byte, n)
	_, err = io.ReadFull(rr.r, rec)
	if err != nil {
		return nil, err
	}
	return rec, nil
}